app/services/courses_query.rb

```diff
@@ -22,6 +23,37 @@ def initialize(params:)
   end
 
   def call
-    @scope.distinct
+    @scope = visa_sponsorship_filter
+    @scope = @scope.distinct
+
+    log_query_info
+
+    @scope
+  end
+
+  def visa_sponsorship_filter
+    return @scope if params[:can_sponsor_visa].blank?
+
+    filter_applied(name: :can_sponsor_visa)
+
+    @scope
+      .where(
+        can_sponsor_student_visa: true
+      )
+      .or(
+        @scope.where(
+          can_sponsor_skilled_worker_visa: true
+        )
+      )
+  end
+
+  private
+
+  def filter_applied(name:)
+    @applied_filters << { name:, value: params[name] }
+  end
```

Comment:
Instead of creating an abstraction here, this could be the following instead:

```ruby
class QueryObject
  def initialize
    self.applied_filters = {}
  end

  def some_filter
    applied_filters[:some_filter] = params[:some_filter]

    @scope
  end

  private

  attr_accessor :applied_filters
end
```

or if you think a filter can be applied multiple times?

```ruby
class QueryObject
  def initialize
    self.applied_filters = []
  end

  def some_filter
    applied_filters << { name: :some_filter, value: params[:some_filter] }

    @scope
  end

  private

  attr_accessor :applied_filters
end
```

Should it?

Reply:
The first one as Hash looks good to me. Done.